### PR TITLE
(LTH-16) Extract facter's dlopen/LoadLibrary wrapper to Leatherman

### PR DESCRIPTION
**Pull request by Magisus**
Extracts facter's dynamic_library classes to Leatherman, including windows and posix variants.

### CODE REVIEW

**Comment by MikaelSmith** [comment]
Needs a rebase.

File: dynamic_library/src/dynamic_library.cc [hunk 1]
```diff
@@ -0,0 +1,58 @@
+#include <leatherman/dynamic_library/dynamic_library.hpp>
+#include <boost/format.hpp>
```
Comment (MikaelSmith) [inline dynamic_library/src/dynamic_library.cc]: Can probably remove `boost/format.hpp`.

File: util/inc/leatherman/util/regex.hpp [hunk 2]
```diff
@@ -7,7 +7,7 @@
 #include <boost/regex.hpp>
 #include <boost/lexical_cast.hpp>
 
-namespace leatherman { namespace file_util {
+namespace leatherman { namespace util {
```
Comment (MikaelSmith) [inline util/inc/leatherman/util/regex.hpp]: Oops. This change would also make sense as a separate `(maint)` commit.

**Comment by MikaelSmith** [comment]
We should add tests to this. They can use dynamic libraries that are expected to be on all systems. `libc` would be guaranteed to be already loaded on POSIX/Mac, @peterhuene any ideas on testing something that's not already loaded?

On Windows, we could try `kernel32.dll` and maybe `gdi32.dll`.

**Comment by Magisus** [comment]
So I'm pretty sure these are failing on Travis and Appveyor because the path to the dummy library is wrong. Where is a good place to put it that is independent of where the test is run, and then how do I reference it?

File: dynamic_library/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,25 @@
+leatherman_dependency(logging)
+leatherman_dependency(util)
+if(WIN32)
+    leatherman_dependency(windows)
+endif()
```
Comment (MikaelSmith) [inline dynamic_library/CMakeLists.txt]: We need to add a dependency on `libdl` for non-OpenBSD linux:
```
else()
    if(NOT CMAKE_SYSTEM_NAME MATCHES "OpenBSD")
        add_leatherman_deps(dl)
    endif()
endif()
```

File: dynamic_library/src/posix/dynamic_library.cc [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        library._first_load = false;
+        return library;
+    }
+
+    bool DynamicLibrary::load(string const& name)
+    {
+        close();
+
+        // Don't actually perform a load to determine if it is already loaded
+#ifdef _AIX
```
Comment (MikaelSmith) [inline dynamic_library/src/posix/dynamic_library.cc]: Need to update based on https://github.com/puppetlabs/facter/commit/858e51a39948a383e0e651d0ca6d3fc1b17edf41.

**Comment by MikaelSmith** [comment]
Needs a rebase.

**Comment by MikaelSmith** [comment]
:+1:

File: dynamic_library/src/windows/dynamic_library.cc [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+#include <boost/nowide/convert.hpp>
+#include <tlhelp32.h>
+
+using namespace std;
+using namespace leatherman::util;
+using namespace leatherman::windows;
+
+namespace leatherman { namespace dynamic_library {
+
+        DynamicLibrary DynamicLibrary::find_by_pattern(std::string const& pattern)
```
Comment (MikaelSmith) [inline dynamic_library/src/windows/dynamic_library.cc]: Indentation here.

File: dynamic_library/src/posix/dynamic_library.cc [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    {
+        if (_handle) {
+            dlclose(_handle);
+            _handle = nullptr;
+        }
+        _name.clear();
+        _first_load = false;
+    }
+
+    void*DynamicLibrary::find_symbol(string const& name, bool throw_if_missing, string const& alias) const
```
Comment (MikaelSmith) [inline dynamic_library/src/posix/dynamic_library.cc]: This compiles, but a space would be more consistent with other styling.

File: dynamic_library/src/dynamic_library.cc [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        return _handle != nullptr;
+    }
+
+    bool DynamicLibrary::first_load() const
+    {
+        return _first_load;
+    }
+
+    string const&DynamicLibrary::name() const
```
Comment (MikaelSmith) [inline dynamic_library/src/dynamic_library.cc]: Same, a space after `&` is more consistent.

File: dynamic_library/tests/dynamic_library_tests.cc [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+        SECTION("should find library function"){
+            REQUIRE(lib.find_symbol("hello"));
+        }
+		
+		lib.close();
+    }
+
+#ifdef _WIN32
+   TEST_CASE("dynamic_library::find_by_pattern", "[dyn-lib]"){
```
Comment (MikaelSmith) [inline dynamic_library/tests/dynamic_library_tests.cc]: Tabs here.

**Comment by MikaelSmith** [comment]
:+1:

**Comment by MikaelSmith** [comment]
Not sure how to add more testing coverage to this. @peterhuene or @branan any suggestions?

File: dynamic_library/tests/dynamic_library_tests.cc [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        REQUIRE(lib.load(lib_path));
+
+        SECTION("should fail to find nonexistent symbol"){
+            REQUIRE_FALSE(lib.find_symbol("not_here"));
+        }
+
+        SECTION("should find library function"){
+            REQUIRE(lib.find_symbol("hello"));
+        }
+
```
Comment (MikaelSmith) [inline dynamic_library/tests/dynamic_library_tests.cc]: We should add a test exercising find_symbol with `throw_if_missing=true`; and one that exercises alias with `alias=goodbye` on libtest1.

File: dynamic_library/tests/dynamic_library_tests.cc [hunk 10]
```diff
@@ -0,0 +1,78 @@
+#include <catch.hpp>
+#include <leatherman/dynamic_library/dynamic_library.hpp>
+
+namespace leatherman { namespace dynamic_library {
+
+#ifdef _WIN32
+    std::string const lib_path = "libtest.so";
+    std::string const lib_path2 = "libtest1.so";
+#else
+    std::string const lib_path = "../lib/libtest.so";
```
Comment (MikaelSmith) [inline dynamic_library/tests/dynamic_library_tests.cc]: Also talked about using a fixture to give these an absolute path.

File: dynamic_library/tests/dynamic_library_tests.cc [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    std::string const lib_path2 = "../lib/libtest1.so";
+#endif
+
+    TEST_CASE("dynamic_library::load and dynamic_library::close", "[dyn-lib]"){
```
Comment (MikaelSmith) [inline dynamic_library/tests/dynamic_library_tests.cc]: Test loading a library that doesn't exist.

File: dynamic_library/tests/dynamic_library_tests.cc [hunk 12]
```diff
@@ -0,0 +1,76 @@
+#include <catch.hpp>
+#include <leatherman/dynamic_library/dynamic_library.hpp>
+#include "fixtures.hpp"
+
+namespace leatherman { namespace dynamic_library {
+
+
+std::string const lib_path = TEST_LIB_DIRECTORY + std::string("/libtest.so");
```
Comment (MikaelSmith) [inline dynamic_library/tests/dynamic_library_tests.cc]: I'd expect this to be indented since it's in a namespace.

**Pull request merged**